some attributes dont work on `typedef`s

```
a.c:1:24: warning: attribute 'sysv_abi' ignored on types [-Wignored-attributes]
typedef __attribute__((sysv_abi)) void (*foo)(void);
                       ^
```

--- a/src/aro/Attribute.zig
+++ b/src/aro/Attribute.zig
@@ -847,7 +847,21 @@ pub fn applyVariableAttributes(p: *Parser, qt: QualType, attr_buf_start: usize, 
         } else {
             try p.attr_application_buf.append(gpa, attr);
         },
-        .calling_convention => try applyCallingConvention(attr, p, tok, base_qt),
+        .calling_convention => try applyKeywordCallingConvention(attr, p, tok, base_qt),
+
+        .fastcall,
+        .stdcall,
+        .thiscall,
+        .vectorcall,
+        .cdecl,
+        .pcs,
+        .riscv_vector_cc,
+        .aarch64_sve_pcs,
+        .aarch64_vector_pcs,
+        .sysv_abi,
+        .ms_abi,
+        => try applyGnuAttrCallingConvention(attr, p, tok, base_qt),
+
         .alloc_size,
         .copy,
         .tls_model,
@@ -902,7 +916,21 @@ pub fn applyTypeAttributes(p: *Parser, qt: QualType, attr_buf_start: usize, diag
             } else {
                 try p.err(tok, .designated_init_invalid, .{});
             },
-            .calling_convention => try applyCallingConvention(attr, p, tok, base_qt),
+            .calling_convention => try applyKeywordCallingConvention(attr, p, tok, base_qt),
+
+            .fastcall,
+            .stdcall,
+            .thiscall,
+            .vectorcall,
+            .cdecl,
+            .pcs,
+            .riscv_vector_cc,
+            .aarch64_sve_pcs,
+            .aarch64_vector_pcs,
+            .sysv_abi,
+            .ms_abi,
+            => try applyGnuAttrCallingConvention(attr, p, tok, base_qt),
+
             .alloc_size,
             .copy,
             .scalar_storage_order,
@@ -958,97 +986,21 @@ pub fn applyFunctionAttributes(p: *Parser, qt: QualType, attr_buf_start: usize) 
         },
         .aligned => try attr.applyAligned(p, base_qt, null),
         .format => try attr.applyFormat(p, base_qt),
-        .calling_convention => try applyCallingConvention(attr, p, tok, base_qt),
-        .fastcall => if (p.comp.target.cpu.arch == .x86) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .fastcall } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"fastcall"});
-        },
-        .stdcall => if (p.comp.target.cpu.arch == .x86) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .stdcall } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"stdcall"});
-        },
-        .thiscall => if (p.comp.target.cpu.arch == .x86) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .thiscall } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"thiscall"});
-        },
-        .vectorcall => if (p.comp.target.cpu.arch == .x86 or p.comp.target.cpu.arch.isAARCH64()) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .vectorcall } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"vectorcall"});
-        },
-        .cdecl => {},
-        .pcs => if (p.comp.target.cpu.arch.isArm()) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = switch (attr.args.pcs.kind) {
-                    .aapcs => .arm_aapcs,
-                    .@"aapcs-vfp" => .arm_aapcs_vfp,
-                } } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"pcs"});
-        },
-        .riscv_vector_cc => if (p.comp.target.cpu.arch.isRISCV()) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .riscv_vector } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"pcs"});
-        },
-        .aarch64_sve_pcs => if (p.comp.target.cpu.arch.isAARCH64()) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .aarch64_sve_pcs } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"pcs"});
-        },
-        .aarch64_vector_pcs => if (p.comp.target.cpu.arch.isAARCH64()) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .aarch64_vector_pcs } },
-                .syntax = attr.syntax,
-            });
-        } else {
-            try p.err(tok, .callconv_not_supported, .{"pcs"});
-        },
-        .sysv_abi => if (p.comp.target.cpu.arch == .x86_64 and p.comp.target.os.tag == .windows) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .x86_64_sysv } },
-                .syntax = attr.syntax,
-            });
-        },
-        .ms_abi => if (p.comp.target.cpu.arch == .x86_64 and p.comp.target.os.tag != .windows) {
-            try p.attr_application_buf.append(gpa, .{
-                .tag = .calling_convention,
-                .args = .{ .calling_convention = .{ .cc = .x86_64_win } },
-                .syntax = attr.syntax,
-            });
-        },
+        .calling_convention => try applyKeywordCallingConvention(attr, p, tok, base_qt),
+
+        .fastcall,
+        .stdcall,
+        .thiscall,
+        .vectorcall,
+        .cdecl,
+        .pcs,
+        .riscv_vector_cc,
+        .aarch64_sve_pcs,
+        .aarch64_vector_pcs,
+        .sysv_abi,
+        .ms_abi,
+        => try applyGnuAttrCallingConvention(attr, p, tok, base_qt),
+
         .malloc => {
             if (base_qt.get(p.comp, .func).?.return_type.isPointer(p.comp)) {
                 try p.attr_application_buf.append(gpa, attr);
@@ -1261,9 +1213,111 @@ fn applyFormat(attr: Attribute, p: *Parser, qt: QualType) !void {
     try p.attr_application_buf.append(p.comp.gpa, attr);
 }
 
-fn applyCallingConvention(attr: Attribute, p: *Parser, tok: TokenIndex, qt: QualType) !void {
-    if (!qt.is(p.comp, .func)) {
-        return p.err(tok, .callconv_non_func, .{ p.tok_ids[tok].symbol(), qt });
+/// These come from GNU attributes like __attribute__((sysv_abi))
+fn applyGnuAttrCallingConvention(attr: Attribute, p: *Parser, tok: TokenIndex, qt: QualType) !void {
+    if (!qt.isCallable(p.comp)) {
+        return p.err(tok, .callconv_non_func, .{ p.tokSlice(tok), qt });
+    }
+    const gpa = p.comp.gpa;
+    switch (attr.tag) {
+        .fastcall => if (p.comp.target.cpu.arch == .x86) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .fastcall } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"fastcall"});
+        },
+        .stdcall => if (p.comp.target.cpu.arch == .x86) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .stdcall } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"stdcall"});
+        },
+        .thiscall => if (p.comp.target.cpu.arch == .x86) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .thiscall } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"thiscall"});
+        },
+        .vectorcall => if (p.comp.target.cpu.arch == .x86 or p.comp.target.cpu.arch.isAARCH64()) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .vectorcall } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"vectorcall"});
+        },
+        .cdecl => {},
+        .pcs => if (p.comp.target.cpu.arch.isArm()) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = switch (attr.args.pcs.kind) {
+                    .aapcs => .arm_aapcs,
+                    .@"aapcs-vfp" => .arm_aapcs_vfp,
+                } } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"pcs"});
+        },
+        .riscv_vector_cc => if (p.comp.target.cpu.arch.isRISCV()) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .riscv_vector } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"pcs"});
+        },
+        .aarch64_sve_pcs => if (p.comp.target.cpu.arch.isAARCH64()) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .aarch64_sve_pcs } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"pcs"});
+        },
+        .aarch64_vector_pcs => if (p.comp.target.cpu.arch.isAARCH64()) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .aarch64_vector_pcs } },
+                .syntax = attr.syntax,
+            });
+        } else {
+            try p.err(tok, .callconv_not_supported, .{"pcs"});
+        },
+        .sysv_abi => if (p.comp.target.cpu.arch == .x86_64 and p.comp.target.os.tag == .windows) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .x86_64_sysv } },
+                .syntax = attr.syntax,
+            });
+        },
+        .ms_abi => if (p.comp.target.cpu.arch == .x86_64 and p.comp.target.os.tag != .windows) {
+            try p.attr_application_buf.append(gpa, .{
+                .tag = .calling_convention,
+                .args = .{ .calling_convention = .{ .cc = .x86_64_win } },
+                .syntax = attr.syntax,
+            });
+        },
+        else => unreachable,
+    }
+}
+
+/// These come from explicit MSVC keywords like __stdcall, __fastcall, etc
+fn applyKeywordCallingConvention(attr: Attribute, p: *Parser, tok: TokenIndex, qt: QualType) !void {
+    if (!qt.isCallable(p.comp)) {
+        return p.err(tok, .callconv_non_func, .{ p.tokSlice(tok), qt });
     }
     switch (attr.args.calling_convention.cc) {
         .c => {},

--- a/src/aro/TypeStore.zig
+++ b/src/aro/TypeStore.zig
@@ -1013,6 +1013,15 @@ pub const QualType = packed struct(u32) {
         return qt.scalarKind(comp).isPointer();
     }
 
+    /// Function or function pointer
+    pub fn isCallable(qt: QualType, comp: *const Compilation) bool {
+        return switch (qt.base(comp).type) {
+            .func => true,
+            .pointer => |ptr| ptr.child.is(comp, .func),
+            else => false,
+        };
+    }
+
     pub fn eqlQualified(a_qt: QualType, b_qt: QualType, comp: *const Compilation) bool {
         if (a_qt.@"const" != b_qt.@"const") return false;
         if (a_qt.@"volatile" != b_qt.@"volatile") return false;

--- a/test/cases/calling convention.c
+++ b/test/cases/calling convention.c
@@ -1,0 +1,13 @@
+__attribute__((sysv_abi)) void sysv_func(void);
+
+__attribute__((sysv_abi)) int x;
+
+
+typedef __attribute__((sysv_abi)) void (*fn_ptr_t)(void);
+void foo(__attribute__((sysv_abi)) void (*p)(void)) {
+	__attribute__((sysv_abi)) void (*fn_ptr)(void);
+}
+
+
+#define EXPECTED_ERRORS "calling convention.c:3:16: warning: 'sysv_abi' only applies to function types; type here is 'int' [-Wignored-attributes]" \
+

--- a/test/cases/msvc attribute keywords.c
+++ b/test/cases/msvc attribute keywords.c
@@ -16,4 +16,4 @@ unsigned long __int64 l;
     ".c:8:13: warning: '__stdcall' only applies to function types; type here is 'int *' [-Wignored-attributes]" \
     ".c:8:32: error: expected ']', found '__cdecl'" \
     ".c:8:31: note: to match this '{'" \
-    ".c:10:17: warning: '__cdecl' only applies to function types; type here is 'void (*)(void)' [-Wignored-attributes]" \
+

